### PR TITLE
Fixes for logging torchmetrics with `sync_dist!=None`  

### LINKED ISSUE #15588: Raise an error if sync_dist is set when logging torchmetrics
# Proposal

Re-do #13903

---

### Context

        > When using torchmetrics, it will sync whether you want it or not.

True, sorry, I noticed this later. We can make it `None` by default.

> [One can chose whether it happens additionally on step or not](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html#torchmetrics.Metric).

we can determine that based on `on_step` and `on_epoch` if `sync_dist!=None`:
- If either just `on_step=True` or just `on_epoch=True`, set the flag inside `torchmetrics` accordingly
- If `on_step=True` and `on_epoch=True`:
  - Set both the flags
  - Set only `sync_on_compute` to align it with how we handle scalar metrics.
  - Raise an error to ask user to set it by themselves, since we can't determine if they want to sync `on_step`, `on_epoch` or on both.

> perhaps some of these checks are already done internally, but I wouldn't know where to look.

if you are talking about:
- in case torchmetrics are logged, what happens when `sync_dist=True`? [It is ignored](https://github.com/Lightning-AI/lightning/blob/0883971ccb973be3f8d0495f442f80f253fd45bb/src/pytorch_lightning/trainer/connectors/logger_connector/result.py#L260-L267)
- what happens on step level if `sync_dist=True`. [Nothing](https://github.com/Lightning-AI/lightning/blob/0883971ccb973be3f8d0495f442f80f253fd45bb/src/pytorch_lightning/trainer/connectors/logger_connector/result.py#L525-L545).

_Originally posted by @rohitgr7 in https://github.com/Lightning-AI/lightning/issues/13903#issuecomment-1206238291_
      
      

cc @borda @carmocca @Blaizzy

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?

Fixes #15588 

True, sorry, I noticed this later. We can make it None by default.

[One can chose whether it happens additionally on step or not](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html#torchmetrics.Metric).

we can determine that based on on_step and on_epoch if sync_dist!=None:

If either just `on_step=True` or just `on_epoch=True`, 

- set the flag inside torchmetrics accordingly

If `on_step=True` and `on_epoch=True`:
- Set both the flags
- Set only sync_on_compute to align it with how we handle scalar metrics.
- Raise an error to ask user to set it by themselves, since we can't determine if they want to sync on_step, on_epoch or on both.

> perhaps some of these checks are already done internally, but I wouldn't know where to look.

if you are talking about:
- in case torchmetrics are logged, what happens when sync_dist=True? [It is ignored](https://github.com/Lightning-AI/lightning/blob/0883971ccb973be3f8d0495f442f80f253fd45bb/src/pytorch_lightning/trainer/connectors/logger_connector/result.py#L260-L267)
- what happens on step level if sync_dist=True. [Nothing](https://github.com/Lightning-AI/lightning/blob/0883971ccb973be3f8d0495f442f80f253fd45bb/src/pytorch_lightning/trainer/connectors/logger_connector/result.py#L525-L545).

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by shenoynikhil** [comment]
@carmocca @awaelchli Can you guys have a look? I've given a first pass over it, but I am not super sure about it.

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ class _METRICS(TypedDict):
 @dataclass
 class _Sync:
     fn: Optional[Callable] = None
-    _should: bool = False
+    _should: Optional[bool] = None
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:51]: Why did you change the type here?

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def __post_init__(self) -> None:
         self._generate_sync_fn()
 
     @property
-    def should(self) -> bool:
+    def should(self) -> Optional[bool]:
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:60]: Same question

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 3]
```diff
@@ -357,6 +357,24 @@ def log(
         rank_zero_only: bool = False,
     ) -> None:
         """See :meth:`~lightning.pytorch.core.module.LightningModule.log`"""
+        # if value is a Metric and sync_dist != None, then
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:360]: These comments just duplicate what was written in code. Comments should describe the "why", not the "what"

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 4]
```diff
@@ -357,6 +357,24 @@ def log(
...
+        if isinstance(value, Metric) and sync_dist is not None:
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:361]: I wouldn't run this logic if `sync_dist = False`. In that case let the metric be run with the configuration it has already

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 5]
```diff
@@ -348,7 +348,7 @@ def log(
         on_epoch: bool = True,
         reduce_fx: Callable = torch.mean,
         enable_graph: bool = False,
-        sync_dist: bool = False,
+        sync_dist: Optional[bool] = None,
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:351]: If we don't deal with the "False" case, then "False" and "None" are equivalent for this logic so you wouldn't need to change this type

**Comment by carmocca** [comment]
Sorry that reviews are slow. We are focused on the 2.0 release, which will land soon.

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py (hunk 5 above)
Comment (shenoynikhil) [reply src/lightning/pytorch/trainer/connectors/logger_connector/result.py]: That makes sense. I made changes to the default types of `should` in `_Sync` based on this change. Have reverted these changes.

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py (hunk 1 above)
Comment (shenoynikhil) [reply src/lightning/pytorch/trainer/connectors/logger_connector/result.py]: Fixed this.

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py (hunk 2 above)
Comment (shenoynikhil) [reply src/lightning/pytorch/trainer/connectors/logger_connector/result.py]: Fixed this.

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py (hunk 3 above)
Comment (shenoynikhil) [reply src/lightning/pytorch/trainer/connectors/logger_connector/result.py]: Changed this. Can you maybe check now?

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py (hunk 4 above)
Comment (shenoynikhil) [reply src/lightning/pytorch/trainer/connectors/logger_connector/result.py]: Makes sense, change it.

**Comment by Borda** [comment]
@shenoynikhil nice job, is this still wip?

**Comment by shenoynikhil** [comment]
> @shenoynikhil nice job, is this still wip?

Hey, was waiting for @carmocca's response.

**Comment by shenoynikhil** [comment]
Will also have to update a few things, I just noticed that more tests are failing now as compared to before.

File: tests/tests_pytorch/trainer/logging_/test_logger_connector.py [hunk 6]
```diff
@@ -644,3 +644,23 @@ def test_result_collection_no_batch_size_extraction():
     assert results["training_step.epoch_log_val"].value == log_val * batch_size
     assert results["training_step.epoch_log_val"].cumulated_batch_size == batch_size
     assert results["training_step.epoch_sum_log_val"].value == log_val
+
+
+def test_sync_dist_with_torchmetrics():
+    acc = Accuracy(task="multiclass", num_classes=4, top_k=1)
```
Comment (carmocca) [inline tests/tests_pytorch/trainer/logging_/test_logger_connector.py:650]: This is failing in CI when the minimum torchmetrics version is installed. cc @Borda

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 7]
```diff
@@ -362,6 +362,23 @@ def log(
...
+        # based on the on_step and on_epoch arguments.
+        if isinstance(value, Metric) and sync_dist:
+            if on_step or on_epoch:
+                value.sync_on_compute = sync_dist
+                value.dist_sync_on_step = on_step
+
+            # All 3 (on_step, on_epoch and sync_dist) should not be set to True
+            if on_step and on_epoch and sync_dist:
+                raise MisconfigurationException(
+                    "Setting self.log(.., on_step=True, on_epoch=True, sync_dist=True)"
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:376]: Can you specify the key: "name" and clarify in the message that this is for logging torchmetrics.Metrics specifically?

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 8]
```diff
@@ -362,6 +362,23 @@ def log(
...
+        # set `sync_on_compute` and `dist_sync_on_step` properties
+        # based on the on_step and on_epoch arguments.
+        if isinstance(value, Metric) and sync_dist:
+            if on_step or on_epoch:
+                value.sync_on_compute = sync_dist
+                value.dist_sync_on_step = on_step
+
+            # All 3 (on_step, on_epoch and sync_dist) should not be set to True
+            if on_step and on_epoch and sync_dist:
+                raise MisconfigurationException(
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:375]: The exception should be raised before the `if` above is checked, otherwise you are running it when you will fail anyways

File: src/lightning/pytorch/trainer/connectors/logger_connector/result.py [hunk 9]
```diff
@@ -362,6 +362,23 @@ def log(
...
+        # For torchmetric.Metric, if sync_dist is True,
+        # set `sync_on_compute` and `dist_sync_on_step` properties
+        # based on the on_step and on_epoch arguments.
+        if isinstance(value, Metric) and sync_dist:
+            if on_step or on_epoch:
+                value.sync_on_compute = sync_dist
+                value.dist_sync_on_step = on_step
+
+            # All 3 (on_step, on_epoch and sync_dist) should not be set to True
+            if on_step and on_epoch and sync_dist:
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/logger_connector/result.py:374]: `sync_dist` is always True as the `if` statement already checks that in L368

**Review by carmocca — COMMENTED** [review]
Sorry, I missed this PR! Thank you for your patience.
